### ISSUE TITLE
Error while converting ONNX to TRT

### ISSUE THREAD

**Issue by aeryncaen** [AUTOMATIC1111/stable-diffusion-webui-tensorrt#24]
```
$ "/home/ubuntu/stable-diffusion-webui/extensions/stable-diffusion-webui-tensorrt/TensorRT-8.6.1.6/bin/trtexec" --onnx="/home/ubuntu/stable-diffusion-webui/models/Unet-onnx/woopwoop-general-2.0.onnx" --saveEngine="/home/ubuntu/stable-diffusion-webui/models/Unet-trt/woopwoop-general-2.0.trt" --minShapes=x:2x4x64x64,context:2x77x768,timesteps:2 --maxShapes=x:32x4x256x256,context:32x770x768,timesteps:32
...
[05/30/2023-17:02:44] [I] TensorRT version: 8.6.1
[05/30/2023-17:02:44] [I] Loading standard plugins
[05/30/2023-17:02:44] [I] [TRT] [MemUsageChange] Init CUDA: CPU +152, GPU +0, now: CPU 158, GPU 5415 (MiB)
[05/30/2023-17:02:52] [I] [TRT] [MemUsageChange] Init builder kernel library: CPU +2478, GPU +756, now: CPU 2712, GPU 6171 (MiB)
[05/30/2023-17:02:52] [W] [TRT] CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage and speed up TensorRT initialization. See "Lazy Loading" section of CUDA documentation https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#lazy-loading
[05/30/2023-17:02:52] [I] Start parsing network model.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:604] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 1721624086
[05/30/2023-17:02:53] [I] [TRT] ----------------------------------------------------------------
[05/30/2023-17:02:53] [I] [TRT] Input filename:   /home/ubuntu/stable-diffusion-webui/models/Unet-onnx/woopwoop-general-2.0.onnx
[05/30/2023-17:02:53] [I] [TRT] ONNX IR version:  0.0.8
[05/30/2023-17:02:53] [I] [TRT] Opset version:    17
[05/30/2023-17:02:53] [I] [TRT] Producer name:    pytorch
[05/30/2023-17:02:53] [I] [TRT] Producer version: 2.0.1
[05/30/2023-17:02:53] [I] [TRT] Domain:
[05/30/2023-17:02:53] [I] [TRT] Model version:    0
[05/30/2023-17:02:53] [I] [TRT] Doc string:
[05/30/2023-17:02:53] [I] [TRT] ----------------------------------------------------------------
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:604] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 1721624086
[05/30/2023-17:02:55] [W] [TRT] onnx2trt_utils.cpp:374: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[05/30/2023-17:02:57] [I] Finished parsing network model. Parse time: 4.66374
[05/30/2023-17:02:57] [I] [TRT] Graph optimization time: 0.317878 seconds.
[05/30/2023-17:02:57] [I] [TRT] [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +0, GPU +256, now: CPU 6403, GPU 39595 (MiB)
[05/30/2023-17:02:57] [I] [TRT] [MemUsageChange] Init cuDNN: CPU +0, GPU +264, now: CPU 6403, GPU 39859 (MiB)
[05/30/2023-17:02:57] [W] [TRT] TensorRT was linked against cuDNN 8.9.0 but loaded cuDNN 8.7.0
[05/30/2023-17:02:57] [I] [TRT] Local timing cache in use. Profiling results in this builder pass will not be stored.
[05/30/2023-17:02:57] [E] Error[4]: kOPT values for profile 0 violate shape constraints: /input_blocks.1/input_blocks.1.1/transformer_blocks.0/attn1/Einsum_output_0: tensor volume exceeds (2^31)-1, dimensions are [(* 8 batch_size),E0,E0] where E0=(* height width) Volume exceeds 2^31-1.
[05/30/2023-17:02:58] [E] Error[4]: [shapeCompiler.cpp::evaluateShapeChecks::1276] Error Code 4: Internal Error (kOPT values for profile 0 violate shape constraints: /input_blocks.1/input_blocks.1.1/transformer_blocks.0/attn1/Einsum_output_0: tensor volume exceeds (2^31)-1, dimensions are [(* 8 batch_size),E0,E0] where E0=(* height width) Volume exceeds 2^31-1.)
[05/30/2023-17:02:58] [E] Engine could not be created from network
[05/30/2023-17:02:58] [E] Building engine failed
[05/30/2023-17:02:58] [E] Failed to create engine from model or file.
[05/30/2023-17:02:58] [E] Engine set up failed
&&&& FAILED TensorRT.trtexec [TensorRT v8601] # /home/ubuntu/stable-diffusion-webui/extensions/stable-diffusion-webui-tensorrt/TensorRT-8.6.1.6/bin/trtexec --onnx=/home/ubuntu/stable-diffusion-webui/models/Unet-onnx/woopwoop-general-2.0.onnx --saveEngine=/home/ubuntu/stable-diffusion-webui/models/Unet-trt/woopwoop-general-2.0.trt --minShapes=x:2x4x64x64,context:2x77x768,timesteps:2 --maxShapes=x:32x4x256x256,context:32x770x768,timesteps:32
```

**Comment by darongma** [comment]
You probably set the max width and max height too high.

for some reason, i am only able to set the max width and height to both 832, anything higher will fail.

i am also trying to find out what can i do to increase the max height and width.

**Comment by jebarpg** [comment]
I did a bunch of testing and found out how to figure out the max combinations of heights with widths. So like it says the volume is to high, I figured out that the max size of the --maxShapes can only total to at highest 92160... that is you take 2\*4\*width\*height has to be equal to or less than 92160, for instance you can have the following:
Maximum width: 1280
Maximum height: 576
which produces in the command: --maxShapes==x:2x4x72x160,...
2x4x72x160 = 92160
this works and runs successfully.

and say we do the following:
Maximum width: 1344 <--- this is the next step up that the width can be.
Maximum height: 576
we get --maxShapes=x:2x4x72x168
2x4x72x168 = 96768 which is over the limit of 92160

one more example for good measures, I'm just going to flip the width and height
Maximum width: 576  
Maximum height: 1344 
we get --maxShapes=2x4x168x72
2x4x168x72 = 96768 which is over the limit of 92160

so basically what we need for the UI is to have a limiter for people when they are sliding the max width and the max height bar. then this will prevent the error from every happening. outside of someone manually running a command line with custom modifications.

I'll see if I can add this into the UI. 

wish me luck lol

**Comment by darongma** [comment]
> [libprotobuf WARNING google/protobuf/io/coded_stream.cc:604] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
> [libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 1721624086

Thanks for the valuable observation. I am now able to get the dimension I want not exceeding the max shape size of 92160.

Thanks!!!

**Comment by darongma** [comment]
so, in other words, according to this formula:  x_max = f"{max_bs * 2}x4x{max_height // 8}x{max_width // 8}"

batch size * height * width / 8 <= 92160 (Conversion Will Succeed)
batch size * height * width / 8 > 92160 (Conversion Will Fail)

I am able to convert with batch size 1, max width 768, max height 960 since 1x768x960/8=92160

**Comment by darongma** [comment]
here is a list of useful dimensions that should work with batch size 1

512x1440
576x1280
640x1152
704x1024
768x960
832x832

**Comment by jebarpg** [comment]
Yea we actually made a UI update today, with the formula so you can check right on the page if you go over the allotted amount. Waiting for a PR to go through. We also added batch processesing which will create all the onnx models from your stable diffusion directory and also all the trt files with the set parameters. I also added an optional check box for adding the shape of the trt model to the end of the models file name, so you'll see model_name_batch_sizeXmax_widthXmax_height.trt

**Comment by jebarpg** [comment]
I also have some recommendations in the UI for people. I recommend only using a max batch size of 2 because 3 your max image size has one dimension less than 512.. which isn't very useful.

**Comment by jebarpg** [comment]
You can read my full explanation here

https://github.com/AUTOMATIC1111/stable-diffusion-webui-tensorrt/pull/32

**Comment by darongma** [comment]
> You can read my full explanation here
> 
> #32

Thank you so much !!!!!

**Comment by jebarpg** [comment]
I made a fix for all these issues. You can check out my fork with the changes here: https://github.com/jebarpg/stable-diffusion-webui-tensorrt
I did all the manual testing and discovered the limits of all the shapes you can create with max width, height and batch sizes. The best batch size and max width and height I have found is bs: 7 maxW: 512 maxH: 512. You can max out the max tokens it has no effect on the shape size limit only the max batch size, max width, and max height have any effect. I also discovered a base number per every batch size from 1 to 11 which lets you know how far you can slide the max width and height. You will get a red label signaling that you are over the limit and green otherwise. I've created a pull request so hopefully it gets integrated in. Also you can do batch processing instead of just one model at a time. for both the onnx files and trt files. NOTE that your settings for max width height batch size tokens etc will apply to the entire batch. Let me know what you all think.
